transports/jms/src/main/java/org/mule/transport/jms/JmsMessageDispatcher.java

```diff
@@ -95,186 +98,310 @@ protected boolean isDisableTemporaryDestinations()
+                    closeSession(session, transacted);
                 }
             }
 
-            // If the session is from the current transaction, it is up to the
-            // transaction to close it.
-            if (session != null && !cached && !transacted)
+            @Override
+            public void onFailure(Exception exception)
             {
-                connector.closeQuietly(session);
+                try
+                {
+                    completionHandler.onFailure(exception);
+                }
+                finally
+                {
+                    closeConsumer(session, consumer, replyTo);
+                    closeSession(session, transacted);
+                }
             }
+        }));
+        connector.getJmsSupport().send(producer, jmsMessage, persistent, priority, ttl, topic, endpoint);
+        connector.getResponseTimeoutTimer().schedule(closeConsumerTask, endpoint.getResponseTimeout());
+    }
+
+    private MessageProducer createProducer(Session session, boolean topic) throws JMSException
+    {
+        final Destination dest = connector.getJmsSupport().createDestination(session, endpoint);
+        return connector.getJmsSupport().createProducer(session, dest, topic);
+    }
+
+    private Message getJmsMessagePayload(MuleEvent event) throws DispatchException
+    {
+        Object message = event.getMessage().getPayload();
+        if (!(message instanceof Message))
+        {
+            throw new DispatchException(
+                    JmsMessages.checkTransformer("JMS message", message.getClass(), connector.getName()),
+                    event, getEndpoint());
+        }
+        return (Message) message;
+    }
+
+    private boolean isUseReplyToDestination(MuleEvent event, boolean doSend, boolean transacted)
+    {
+        return returnResponse(event, doSend) && !transacted;
+    }
+
+    private boolean isTransacted()
+    {
+        Transaction muleTx = TransactionCoordination.getInstance().getTransaction();
+        return (muleTx != null && muleTx.hasResource(connector.getConnection()) || endpoint.getTransactionConfig().isTransacted());
+    }
+
+    private void closeConsumer(Session session, MessageConsumer consumer, Destination replyTo)
+    {
+        connector.closeQuietly(consumer);
+
+        // TODO AP check if TopicResolver is to be utilized for temp destinations as well
+        if (replyTo != null && (replyTo instanceof TemporaryQueue || replyTo instanceof TemporaryTopic))
+        {
+            if (replyTo instanceof TemporaryQueue)
+            {
+                connector.closeQuietly((TemporaryQueue) replyTo);
+            }
+            else
+            {
+                // hope there are no more non-standard tricks from JMS vendors
+                // here ;)
+                connector.closeQuietly((TemporaryTopic) replyTo);
+            }
+        }
+    }
+
+    private void closeSession(Session session, boolean transacted)
```

Comment:
Can you remove the transacted parameter and use the isTransacted method instead?